FE-484 | Add queryTimeout to Client (#264)
fauna/faunadb-js: README.md

[See the JSDocs](https://fauna.github.com/faunadb-js/PageHelper.html) for
more information on the pagination helper.

#### Timeouts

The client can be configured to handle timeouts in two different ways:

1. Add a `timeout` field to the `options` block when instantiating the client
2. By setting a `queryTimeout` on the client (or passing the value to the client's `.query()` method directly)

> The first option (i.e. `timeout`) represents a HTTP timeout on the client side. Represented in milliseconds, the client will wait the specified period before timing out if it has yet to receive data.

Reviewer: ```suggestion
The first option (i.e. `timeout`) represents a HTTP timeout on the client side. Represented in milliseconds, the client will wait the specified period before timing out if it has yet to receive a response.
```

```javascript
const client = new faunadb.Client({
  secret: 'YOUR_FAUNADB_SECRET',
  timeout: 100,
})
```

> On the other hand, using the client's `queryTimeout` dictates how long FaunaDB will process the request on the server-side before timing out if it hasn't finished it's lookup and/or returned data. This can be done in two different ways:

Reviewer: ```suggestion
On the other hand, using the client's `queryTimeout` dictates how long FaunaDB will process the request on the server before timing out if it hasn't finished running the operation. This can be done in two different ways:
```

more information on the pagination helper.

#### Timeouts

The client can be configured to handle timeouts in two different ways:

1. Add a `timeout` field to the `options` block when instantiating the client
2. By setting a `queryTimeout` on the client (or passing the value to the client's `.query()` method directly)

The first option (i.e. `timeout`) represents a HTTP timeout on the client side. Represented in milliseconds, the client will wait the specified period before timing out if it has yet to receive a response.

```javascript
const client = new faunadb.Client({
  secret: 'YOUR_FAUNADB_SECRET',
  timeout: 100,
})
```

On the other hand, using the client's `queryTimeout` dictates how long FaunaDB will process the request on the server before timing out if it hasn't finished running the operation. This can be done in three different ways:

```javascript
// 1. Setting the value when instantiating a new client
const client = new faunadb.Client({
  queryTimeout: 2000,
  secret: 'YOUR_FAUNADB_SECRET',
})

// 2. Setting the value on the client instance
const client = new faunadb.Client({
  secret: 'YOUR_FAUNADB_SECRET',
})

> client.queryTimeout(200)

Reviewer 2: usually a I don't like mutate the object state, if I'm using the same client to issue multiple queries and change this property in one part of the code, this will have side effects in other parts of code not related, my opinion is to remove this option and leave it only on constructor and on query method

Reviewer: Personally, I also prefer having immutable clients too, but I think that's relatively uncommon for other similar JS clients. Before this change was the client completely immutable already?

Reviewer 2: besides this setter https://github.com/fauna/faunadb-js/blob/master/src/Client.js#L152 that's meant to be used to sync last txn time (internally only in my option) there are no other setter.

Reviewer: Ok so in that case I'm inclined to not introduce mutability into the client with this change.

Here are two options I can think of:
1. remove the `.queryTimeout()` setter.
2. make the `.queryTimeout()` method return a new client with the set parameter.

Author: @nathaniel-may I discussed this with the team and we are in favor of #1. If users end up asking for something like #2, we can always add it later. What do you think?

Reviewer: Sounds good to me. Thanks for raising the issue, @marrony.
